### DISCUSSION TITLE
#import bevy_pbr::pbr_functions error

### DISCUSSION THREAD

**Discussion by stan4cb** [bevyengine/bevy#8498] [Q&A] [answered]
Hi, I'm trying learn and write a custom shader in wgsl, but I don't want to lose shadows so it seems that I must import 

```
#import bevy_pbr::pbr_types
#import bevy_pbr::pbr_functions
```

and create a config with `pbr_input_new()` to pass into `pbr()`

but the problem is 

`#import bevy_pbr::pbr_functions` results in this error

```
error: no definition in scope for identifier: 'view'
    ┌─ wgsl:317:31
    │
317 │     color = saturation(color, view.color_grading.pre_saturation);
    │                               ^^^^ unknown identifier
```

this snippet seems enough to replicate error

```
#import bevy_pbr::pbr_types
#import bevy_pbr::pbr_functions

struct CustomMaterial {
    color: vec4<f32>,
};

struct FragmentInput {
    #import bevy_pbr::mesh_vertex_output
}

@group(1) @binding(0)
var<uniform> material: CustomMaterial;

@fragment
fn fragment(
    in: FragmentInput
) -> @location(0) vec4<f32> {
    return material.color;
}
```

is this a bug or I am missing something?

**Comment by nicopap** [comment, marked as answer]
You need to add the following imports to your shader:

```
#import bevy_pbr::mesh_bindings
#import bevy_pbr::mesh_view_bindings
#import bevy_pbr::pbr_bindings
```

`view` is a uniform defined in `mesh_view_bindings`. It is not imported by either `pbr_types` or `pbr_functions`, therefore if you don't import `mesh_view_bindings`, it won't be available for them.

I expect that other functions in `pbr_functions` require other bindings defined in the files I listed.

In all fairness, it's a poorly documented side of bevy that is not intuitive.

**Comment by stan4cb** [comment]
at the end all the imports I had to have was:

```
#import bevy_pbr::mesh_bindings
#import bevy_pbr::mesh_view_bindings
#import bevy_pbr::pbr_bindings

#import bevy_pbr::fog
#import bevy_pbr::shadows
#import bevy_pbr::lighting
#import bevy_pbr::pbr_ambient
#import bevy_pbr::clustered_forward

#import bevy_pbr::utils

//#import bevy_pbr::pbr_types
#import bevy_pbr::pbr_functions
```
